### ISSUE TITLE
Cosmic Appreance Settings Crashes

### ISSUE THREAD

**Issue by asosnovsky** [lilyinstarlight/nixos-cosmic#440]
All of the other settings seem to work fine, but for some reason when I go to the "appereance" section it crashes and exits.

Here are the logs I gathered when launching the settings from terminal:

```
 ~ cosmic-settings
   WARN  Key-Binding 'Period' only matched case insensitive for "period"
    at /build/cargo-vendor-dir/cosmic-settings-config-0.1.0/src/shortcuts/sym.rs:26 on main

   WARN  Key-Binding 'Comma' only matched case insensitive for "comma"
    at /build/cargo-vendor-dir/cosmic-settings-config-0.1.0/src/shortcuts/sym.rs:26 on main

   WARN  Key-Binding 'Period' only matched case insensitive for "period"
    at /build/cargo-vendor-dir/cosmic-settings-config-0.1.0/src/shortcuts/sym.rs:26 on main

   WARN  Key-Binding 'Comma' only matched case insensitive for "comma"
    at /build/cargo-vendor-dir/cosmic-settings-config-0.1.0/src/shortcuts/sym.rs:26 on main

  ERROR  shortcuts custom config error: GetKey("custom", Os { code: 2, kind: NotFound, message: "No such file or directory" })
    at /build/cargo-vendor-dir/cosmic-settings-config-0.1.0/src/shortcuts/mod.rs:43 on main

  ERROR  Failed to get the contrast Portal(NotFound("Requested setting not found"))
    at /build/cargo-vendor-dir/libcosmic-0.1.0/src/theme/portal.rs:47 on tokio-runtime-worker

The application panicked (crashed).
Message:  Size not found for icon
Location: /build/cargo-vendor-dir/freedesktop-icons-0.2.6/src/theme/parse.rs:54

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: PoisonError { .. }
Location: /build/cargo-vendor-dir/libcosmic-0.1.0/src/widget/icon/named.rs:78

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: PoisonError { .. }
Location: /build/cargo-vendor-dir/libcosmic-0.1.0/src/icon_theme.rs:17

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: PoisonError { .. }
Location: /build/cargo-vendor-dir/libcosmic-0.1.0/src/icon_theme.rs:17

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

Here is some info:

```
cat flake.lock | jq '.nodes["nixos-cosmic"]'
{
  "inputs": {
    "flake-compat": "flake-compat",
    "nixpkgs": [
      "nixpkgs"
    ],
    "nixpkgs-stable": "nixpkgs-stable",
    "rust-overlay": "rust-overlay"
  },
  "locked": {
    "lastModified": 1729857853,
    "narHash": "sha256-IVaFOTG4i2K0YWKrJui09YCAEWyTSK+zaUTUvj/SbC4=",
    "owner": "lilyinstarlight",
    "repo": "nixos-cosmic",
    "rev": "04408bf4afe2bf2b15227c43914130c8bdf4ed3c",
    "type": "github"
  },
  "original": {
    "owner": "lilyinstarlight",
    "repo": "nixos-cosmic",
    "type": "github"
  }
}
```

```
cat flake.lock | jq '.nodes["nixpkgs"]'
{
  "locked": {
    "lastModified": 1729449015,
    "narHash": "sha256-Gf04dXB0n4q0A9G5nTGH3zuMGr6jtJppqdeljxua1fo=",
    "owner": "NixOS",
    "repo": "nixpkgs",
    "rev": "89172919243df199fe237ba0f776c3e3e3d72367",
    "type": "github"
  },
  "original": {
    "owner": "NixOS",
    "ref": "nixos-24.05",
    "repo": "nixpkgs",
    "type": "github"
  }
}
```

**Comment by lilyinstarlight** [comment]
I cannot replicate this issue at all on latest nixos-unstable or nixos-24.05. Is your config publicly available so I can poke at it?

**Comment by asosnovsky** [comment]
@lilyinstarlight yes it is https://github.com/asosnovsky/nixos-setup/tree/main

This host is the one for my laptop: https://github.com/asosnovsky/nixos-setup/blob/main/hosts/fwbook.nix

The desktop config https://github.com/asosnovsky/nixos-setup/blob/main/modules/nixos/desktop/cosmic.nix

FYI I am running it side by side with Plasma6, and using GDM. Though I did try the cosmic-gretter and got the same results.

**Comment by asosnovsky** [comment]
@lilyinstarlight any updates? I think the root cause must be just that when this `.unwrap` https://github.com/pop-os/libcosmic/blob/36b3cfa13aeac6e5cdf7e76ef826ed7d66aaebfe/src/icon_theme.rs#L17 is called it corrupts the locks which then crashes the whole thing. 

I am guessing that I am missing some default styles for cosmic? Not sure why or how... any chance the log message could at least point me to the file I need to add?

**Comment by lilyinstarlight** [comment]
You don't need to add any files, the module handles everything. I'll try to poke at your config today or tomorrow to see what's wrong with it

**Comment by lilyinstarlight** [comment]
To clarify, the log says even at the start (which I was suspicious of) that you're missing COSMIC schemas. That does not happen in the configurations provided by the module in this repository, so I'm worried something on your end is mucking something up after this module is applied. I'm trying to build your config now and manually inspect the toplevel

Also the application is crashing elsewhere from that line as well. Look at the earlier backtraces, another thread crashed first and the main thread only failed there when a different thread died while holding the lock (and `.unwrap()` doesn't corrupt anything, it specifically calls `abort()` because that function is the programmer's way of saying "if this is an error then kill the application" in Rust)

**Comment by asosnovsky** [comment]
@lilyinstarlight that makes sense. I also suspected something is strange with my gtk theming. I have a feeling that I messed up something when installing a custom theme via KDE plasma a few months back (unfortunately this was not done declaratively).

What's the appropriate way to reset all the changes to my gtk theming? Is there a folder under .config I can clear?

**Comment by Thatoo** [comment]
I should say that, if it has never crash, this setting panel (Desktop > Appearance) takes quiet some time to display compare to all other settings panels. Is it normal?

**Comment by lilyinstarlight** [comment]
> What's the appropriate way to reset all the changes to my gtk theming? Is there a folder under .config I can clear?

Hmmmm spooky

If you're sure there's nothing you need out of these directories, you could `rm -r ~/.config/gtk-* ~/.local/share/{themes,icons}` I suppose. But perhaps it would be safer to make a backup of them (e.g. `mkdir ~/themebackup && mv ~/.config/gtk-* ~/.local/share/{themes,icons} ~/themebackup/`)

**Comment by lilyinstarlight** [comment]
> I should say that, if it has never crash, this setting panel (Desktop > Appearance) takes quiet some time to display compare to all other settings panels. Is it normal?

How much is "quite some time"? (also this is likely unrelated to this issue so if you could open a new issue or just ping me on Matrix if you're not sure it's an issue that would be helpful <3)

**Comment by asosnovsky** [comment]
@lilyinstarlight clearing out my icons folder in share seems to have resolved this. It seems that the icons in this theme https://github.com/vinceliuice/WhiteSur-gtk-theme cause the crash somehow.

I can also confirm what @Thatoo where opening the appreance sections takes a noticable 2-3 seconds before I can change any settings.

Also I noticed that some buttons like "Apply this theme to GNOME apps" just don't click.

**Comment by Elias-Graf** [comment]
> How much is "quite some time"? (also this is likely unrelated to this issue so if you could open a new issue or just ping me on Matrix if you're not sure it's an issue that would be helpful <3)

I used to experience this delay too (about 3–8 seconds to load the page), but I simply assumed, it was some weird stuff with the daemon. Now, as of the latest build, the appearance settings seemed to load like every other page.

**Comment by Thatoo** [comment]
I keep having this delay even after updating right now.

**Comment by Elias-Graf** [comment]
Yup, same, I just checked again after rebooting from yesterday, and the first time I open it, it takes a solid 5 seconds to load the page. After that, it seems to open immediately.

**Comment by Thatoo** [comment]
Since last updates, cosmic appearance settings don't crash and display very fast now. I don't experience any issue anymore. Maybe this issue could be closed now.

**Comment by Elias-Graf** [comment]
Quickly tested it on two machines without issue.

**Comment by asosnovsky** [comment]
I can confirm it’s working fine for me as well!
